### DISCUSSION TITLE
Transaction not rolling back

### DISCUSSION THREAD

**Discussion by Vishwas-verma** [adonisjs/core#2770] [General]
Hey I dont know why my transaction is not rolling back after failing:

**Controller Code-**
```
  public async registerDriver({request, response}: HttpContextContract) {
    const inputData = (await request.validate(DriverSignupValidator)) as DriverCreateDto;
    const userData = inputData.user;
    const trx = await Database.transaction()
    let user: User;
    try {
      user = await userService.createUser(userData, RoleEnum.DRIVER, trx);
      await userService.createDriver(inputData, user.id, trx);
      await user.load('driver');
      await trx.commit();
    } catch (e) {
      logger.error("Transaction Failed ", e);
      await trx.rollback();
      throw e;
    }
    return response.json({
      user: await new UserTransformer().transform(user, ["driver"])
    });
  }
```

**Service Code-**

```
   async createDriver(data: DriverCreateDto, userId: number, trx: TransactionClientContract): Promise<Driver> {
    const driver = await Driver.create({
      ...data,
      userId,
    }, trx)
    await driver.related('identity').create(data.identity);
    await BankDetail.create({...data.bankDetail, driverId: driver.id}, trx);
    await Cab.create({...data.cab, driverId: driver.id}, trx);
    await driver.load('cab');
    await driver.load('bankDetail');
    await driver.load('identity');
    return driver;
  }
---------------------------------------------------------------------------------------------------------------------------------------------------

  async createUser(data: UserCreateDto, role: RoleEnum, trx?: TransactionClientContract): Promise<User> {
    const referralCode = string.generateRandom(7);
    const user = await User.create({...data, referralCode, role}, trx);
    if (!!data.documents) await user.related('documents').createMany(data.documents);
    await user.load('documents');
    return user;
  }

```

If my code fail while creating Cab the transaction doesnt roll back new user ,new driver and bank details are still created.

**Comment by thetutlage** [comment]
You have not shared the code for the `createUser` method. However, in the `createDriver` method, you are not using the `trx` object when calling `await Driver.create()` method

**Reply by Vishwas-verma** [reply]
Hey @thetutlage I have updated the code it still doesn't seems to work.

**Reply by thetutlage** [reply]
It is still using the `trx` incorrect inside the create method. I am sure, the typescript compiler is also shown you an error. So please fix your code, format it properly and then post a new comment with the updated code. 

Here are the docs for the create method https://docs.adonisjs.com/reference/orm/base-model#static-create

**Reply by Vishwas-verma** [reply]
Thanks @thetutlage it worked , the typescript was not giving error in wrong use of ```trx``` which is quite strange.
Though I found a correct way for it which is,
```
    const driver = await Driver.create({
      ...data,
      userId,
    }, {client:trx})
```

**Reply by GnRlLeclerc** [reply]
I spend a couple debugging this too. What I find strange is that :
```
    const driver = await Driver.create({
      ...data,
      userId,
    }, trx)
```
...does indeed work "as expected" when in testing environment using `node ace test`, as per the adonis documentation.

However, when running the backend, it does not work and the following syntax gives the expected result.

```
    const driver = await Driver.create({
      ...data,
      userId,
    }, {client:trx})
```

This was a bit confusing to debug because (in addition to the inconsistent behavior), `useTransaction` directly expects the `trx` object, while other methods expect `{client:trx}`. TypeScript does not detect anything wrong.
